agent/metrics/metrics_test.go

```diff
@@ -0,0 +1,222 @@
+			go func() {
+				defer wg.Done()
+				defer MetricsEngineGlobal.RecordDockerMetric("STOP")()
+				time.Sleep(DockerMetricSleepTime2)
+			}()
+		}
+	}()
+	time.Sleep(2 * time.Second)
+	wg.Wait()
+
+	// This will gather all collected metrics and store them in a MetricFamily list
+	// All metric families can be printed by looping over this variable using
+	// fmt.Println(proto.MarshalTextString(metricFamilies[n])) where n = index
+	metricFamilies, err := MetricsEngineGlobal.Registry.Gather()
+	assert.NoError(t, err)
+
+	// Here we set up the expectations. These are known values which make verfication
+	// easier.
+	expected := make(metricMap)
+	expected["AgentMetrics_DockerAPI_call_count"] = make(map[string][]interface{})
+	expected["AgentMetrics_DockerAPI_call_count"]["CallSTART"] = []interface{}{
+		"COUNTER",
+		20.0,
+	}
+	expected["AgentMetrics_DockerAPI_call_count"]["CallSTOP"] = []interface{}{
+		"COUNTER",
+		20.0,
+	}
+	expected["AgentMetrics_DockerAPI_call_duration"] = make(map[string][]interface{})
+	expected["AgentMetrics_DockerAPI_call_duration"]["CallSTART"] = []interface{}{
+		"GUAGE",
+		0.0,
+	}
+	expected["AgentMetrics_DockerAPI_call_duration"]["CallSTOP"] = []interface{}{
+		"GUAGE",
+		0.0,
+	}
+	expected["AgentMetrics_DockerAPI_duration_seconds"] = make(map[string][]interface{})
+	expected["AgentMetrics_DockerAPI_duration_seconds"]["CallSTART"] = []interface{}{
+		"SUMMARY",
+		3.0,
+	}
+	expected["AgentMetrics_DockerAPI_duration_seconds"]["CallSTOP"] = []interface{}{
+		"SUMMARY",
+		3.0,
+	}
+	// We will do a simple tree search to verify all metrics in metricsFamilies
+	// are as expected
+	assert.True(t, VerifyStats(metricFamilies, expected), "Metrics are not accurate")
+}
+
+// A type for storing a Tree-based map. We map the MetricName to a map of metrics
+// under that name. This second map indexes by MetricLabelName+MetricLabelValue to
+// a slice MetricType and MetricValue.
+//MetricName:metricLabelName+metricLabelValue:[metricType, metricValue]
+type metricMap map[string]map[string][]interface{}
+
+// In order to verify the MetricFamily with the expected metric values, we do a simple
+// tree search to verify that all stats in the MetricFamily coincide with the expected
+// metric values.
+// This method only verifes that all metrics in var metricsReceived are present in
+// var expectedMetrics
+func VerifyStats(metricsReceived []*dto.MetricFamily, expectedMetrics metricMap) bool {
+	var threshhold float64 = 0.1 // Maximum threshhold for two metrics being equal
+	for _, metricFamily := range metricsReceived {
+		if metricList, found := expectedMetrics[metricFamily.GetName()]; found {
+			for _, metric := range metricFamily.GetMetric() {
+				if aMetric, found := metricList[metric.GetLabel()[0].GetName()+metric.GetLabel()[0].GetValue()]; found {
+					metricTypeExpected := string(aMetric[0].(string))
+					metricValExpected := float64(aMetric[1].(float64))
+					switch metricTypeExpected {
+					case "GUAGE":
```

Comment:
This verification loops through received metrics and checks that every one exists in expected metrics (since it's easier to look up through a map). We need a place holder that ignores GUAGE or else it will not verify correctly.